rs/embedders/src/wasmtime_embedder/system_api.rs

```diff
@@ -1049,12 +1058,96 @@ pub fn syscalls<
+    // pages are there on the heap.
+    linker
+        .func_wrap("__", "check_accessed_pages", {
+            move |mut caller: Caller<'_, StoreData>| -> Result<(), _> {
+                with_error_handling(&mut caller, |mut c| {
+                    // Using the libc mincore() API, we can check how many pages are accessed.
+                    // we can give it the start address of the heap and the size of the heap.
+                    // The mincore() API will return a vector of bytes, where each bit represents
+                    // a page in the heap. If the bit is set, the page is in memory, if it is not set,
+                    // the page is not in memory.
+
+                    // If the API type is not replicated execution, return immediately.
+                    if !c.data().system_api().unwrap().is_replicated_message() {
+                        return Ok(());
+                    }
+
+                    // Get memory from caller c.
+                    let memory = c
+                        .get_export(WASM_HEAP_MEMORY_NAME)
+                        .ok_or_else(|| HypervisorError::ToolchainContractViolation {
+                            error: "WebAssembly module must define memory".to_string(),
+                        })
+                        .and_then(|ext| {
+                            ext.into_memory().ok_or_else(|| {
+                                HypervisorError::ToolchainContractViolation {
+                                    error: "export 'memory' is not a memory".to_string(),
+                                }
+                            })
+                        })
+                        .map(|mem| {
+                            // False positive clippy lint.
+                            // Issue: https://github.com/rust-lang/rust-clippy/issues/12856
+                            // Fixed in: https://github.com/rust-lang/rust-clippy/pull/12892
+                            #[allow(clippy::needless_borrows_for_generic_args)]
+                            let (mem, _) = mem.data_and_store_mut(&mut c);
+                            mem
+                        })
+                        .unwrap();
+
+                    // memory is &mut [u8]. cast to *mut c_void.
+                    let heap_start = memory.as_mut_ptr();
+                    let heap_size = memory.len();
+                    let num_pages = heap_size / 4096;
+
+                    let mut vec = vec![0u8; num_pages];
+                    let res = unsafe {
+                        libc::mincore(
+                            heap_start as *mut libc::c_void,
+                            heap_size as libc::size_t,
+                            vec.as_mut_ptr(),
+                        )
+                    };
+                    if res == 0 {
+                        // Count the number of accessed pages.
+                        let accessed_pages = vec.iter().filter(|&&x| x != 0).count();
```

Comment:
Is this actually counting the bits?

Reply:
I guess it's not very clear -- I modified this to check if each byte (representing one page) is 1.